### PR TITLE
ENH pytorch solvers

### PR DESCRIPTION
**Pull request by tomMoral**
Trying to make a pure `pytorch` solver, to avoid callback overhead.

- I renamed the orignal one in `lightning` so we don't lose any code.
- I did a minimal solver based on what we already have. This seems to give faster results but the code in `tf` seems broken right now. Need to investigate this tomorrow.

### CODE REVIEW

File: objective.py [hunk 1]
```diff
@@ -120,14 +122,26 @@ def _model_init_fn():
             is_small_images = self.width < self.image_width_cutout
             if is_resnet and is_small_images:
                 model = remove_initial_downsample(model)
+            return model
```
Comment (zaccharieramzi) [inline objective.py:125]:
```suggestion
	    if torch.cuda.is_available():
	    	model = model.cuda()
            return model
```

File: objective.py [hunk 2]
```diff
@@ -209,6 +225,20 @@ def compute(self, model):
         results["value"] = results["train_loss"]
         return results
 
+    def eval_torch(self, model, dataloader):
+
+        criterion = torch.nn.CrossEntropyLoss()
+        res = {'loss': 0., 'acc': 0, 'n_samples': 0}
+        for X, y in tqdm(dataloader):
+            X, y = X.cuda(), y.cuda()
```
Comment (zaccharieramzi) [inline objective.py:233]:
```suggestion
	    if torch.cuda.is_available():
	            X, y = X.cuda(), y.cuda()
```

File: utils/torch_solver.py [hunk 3]
```diff
@@ -9,18 +10,14 @@
...
-    )
     AugmentedDataset = import_ctx.import_from(
-        'torch_helper', 'AugmentedDataset'
+        'lightning_helper', 'AugmentedDataset'
     )
 
 
 class TorchSolver(BaseSolver):
-    """Torch base solver"""
+    """Pytorch based solver"""
```
Comment (zaccharieramzi) [inline utils/torch_solver.py:20]:
```suggestion
    """Pytorch base solver"""
```

For consistency: basically I am saying that is a "base solver"

File: utils/torch_solver.py [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import os
 import sys
+from tqdm import tqdm
```
Comment (zaccharieramzi) [inline utils/torch_solver.py:3]: Maybe we can move this in the safe import context?

File: utils/torch_solver.py [hunk 5]
```diff
@@ -88,32 +82,38 @@ def set_lr_schedule_and_optimizer(self):
...
 
     @staticmethod
     def get_next(stop_val):
         return stop_val + 1
 
     def run(self, callback):
         # model weight initialization
-        self.model = self.model_init_fn()
+        model = self.model_init_fn()
+        model.cuda()
```
Comment (zaccharieramzi) [inline utils/torch_solver.py:100]:
```suggestion
```

I would put it in the model init fn as suggested above

File: utils/torch_solver.py [hunk 6]
```diff
@@ -88,32 +82,38 @@ def set_lr_schedule_and_optimizer(self):
...
-        # bad interaction with benchopt. Removing this limitation would be
-        # nice to allow multi-GPU training.
-        trainer = Trainer(
-            max_epochs=-1, callbacks=[BenchoptCallback(callback)],
-            accelerator="auto", devices=1
-        )
-        trainer.fit(self.model, train_dataloaders=self.dataloader)
+        while callback(model):
+            for X, y in tqdm(self.dataloader):
+                X, y = X.cuda(), y.cuda()
```
Comment (zaccharieramzi) [inline utils/torch_solver.py:108]:
```suggestion
		if torch.cuda.is_available():
	                X, y = X.cuda(), y.cuda()
```

**Review by zaccharieramzi — APPROVED** [review]
LGTM except for the handling of the cuda stuff

File: utils/torch_solver.py (hunk 4 above)
Comment (tomMoral) [reply utils/torch_solver.py]: yes, we should :)

**Pull request merged**